parser/prism/Translator.cc

```diff
@@ -4789,8 +4823,88 @@ unique_ptr<parser::Node> Translator::translateRescue(pm_begin_node *parentBeginN
             resbodyLoc = core::LocOffsets{resbodyLoc.beginPos(), endPos};
         }
 
-        rescueBodies.emplace_back(
-            make_unique<parser::Resbody>(resbodyLoc, move(exceptionsArray), move(var), move(rescueBody)));
+        bool exceptionsHaveExpr = (exceptionsArray == nullptr || exceptionsArray->hasDesugaredExpr());
+        bool allHaveExpr = exceptionsHaveExpr && hasExpr(var) && hasExpr(rescueBody);
+        if (!directlyDesugar || !allHaveExpr) {
+            auto body = make_unique<parser::Resbody>(resbodyLoc, move(exceptionsArray), move(var), move(rescueBody));
+            allRescueBodiesHaveExpr = false;
+            rescueBodies.emplace_back(move(body));
+        } else {
+            // Build ast::RescueCase expression
+            ast::RescueCase::EXCEPTION_store astExceptions;
+            if (exceptionsArray != nullptr) {
+                auto exceptionsExpr = exceptionsArray->takeDesugaredExpr();
+                if (auto exceptionsArrayExpr = ast::cast_tree<ast::Array>(exceptionsExpr)) {
+                    for (auto &elem : exceptionsArrayExpr->elems) {
```

Comment:
Worth reserving here?

```suggestion
                    astExceptions.reserve(exceptionsArrayExpr->elems.size());
                    
                    for (auto &elem : exceptionsArrayExpr->elems) {
```